File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2020/Y20D12.kt
package org.gristle.pdxpuzzles.advent.y2020

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.enums.Nsew
import org.gristle.pdxpuzzles.utilities.objects.Coord

class Y20D12(input: String) : Day {

    // Holds instruction info
    data class Instruction(val action: Char, val amount: Int)

    // Interface allows use of two different State objects with different values/functions to work with the 
    // solve function.
    interface ShipState {
        val pos: Coord
        fun nextState(instruction: Instruction): ShipState
    }

    // State used for part 1. Holds the current position and direction. 
    data class DirShipState(override val pos: Coord, val dir: Nsew) : ShipState {
        // Provides a new state based on part 1 execution of instructions.
        override fun nextState(instruction: Instruction) = when (instruction.action) {
            'N' -> copy(pos = pos.north(instruction.amount))
            'S' -> copy(pos = pos.south(instruction.amount))
            'E' -> copy(pos = pos.east(instruction.amount))
            'W' -> copy(pos = pos.west(instruction.amount))
            'L' -> copy(dir = dir.multiLeft(instruction.amount / 90))
            'R' -> copy(dir = dir.multiRight(instruction.amount / 90))
            'F' -> copy(pos = pos.move(dir, instruction.amount))
            else -> throw IllegalArgumentException("Invalid Instruction: $instruction")
        }
    }

    // State used for part 2. Holds the current position and waypoint coordinates.
    data class WaypointShipState(override val pos: Coord, val waypoint: Coord) : ShipState {
        // Provides a new state based on part 2 execution of instructions.
        override fun nextState(instruction: Instruction) = when (instruction.action) {
            'N' -> copy(waypoint = waypoint.north(instruction.amount))
            'S' -> copy(waypoint = waypoint.south(instruction.amount))
            'E' -> copy(waypoint = waypoint.east(instruction.amount))
            'W' -> copy(waypoint = waypoint.west(instruction.amount))
            'L' -> copy(waypoint = (0 until (instruction.amount / 90) % 4).fold(waypoint) { acc, _ -> Coord(acc.y, -acc.x) })
            'R' -> copy(waypoint = (0 until (instruction.amount / 90) % 4).fold(waypoint) { acc, _ -> Coord(-acc.y, acc.x) })
            'F' -> copy(pos = (1..instruction.amount).fold(pos) { acc, _ -> acc + waypoint })
            else -> throw IllegalArgumentException("Invalid Instruction: $instruction")
        }
    }

    // parses input into list of Instructions
    private val instructions: List<Instruction> = input
        .lines()
        .map { Instruction(it.first(), it.drop(1).toInt()) }

    // for both parts, start with initial state, execute instructions on each successive state, then take the
    // final location and find the distance from the origin.
    private fun solve(initialShipState: ShipState): Int = instructions
        .fold(initialShipState, ShipState::nextState)
        .pos
        .manhattanDistance()

    override fun part1() = solve(DirShipState(Coord.ORIGIN, Nsew.EAST))

    override fun part2() = solve(WaypointShipState(Coord.ORIGIN, Coord(10, -1)))
}

fun main() = Day.runDay(Y20D12::class)

//    Class creation: 26ms
//    Part 1: 2280 (4ms)
//    Part 2: 38693 (3ms)
//    Total time: 35ms

File: rust/advent/src/bin/2020/y20d12.rs
use advent::utilities::get_input::get_input;
use utilities::{
    enums::cardinals::Cardinal,
    structs::{
        coord::Coord2,
        stopwatch::{ReportDuration, Stopwatch},
    },
};

type Instruction = (char, Output);
type Input = Vec<Instruction>;
type Output = usize;
type Pos = Coord2;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(20, 12).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

#[derive(Debug)]
enum ShipState {
    Dir { pos: Pos, dir: Cardinal },
    Waypoint { pos: Pos, waypoint: Pos },
}

impl ShipState {
    fn next_state(&self, (action, amount): Instruction) -> Self {
        match *self {
            ShipState::Dir { pos, dir } => match action {
                'N' => Self::Dir {
                    pos: pos.move_direction(Cardinal::North, amount as i64).unwrap(),
                    dir,
                },
                'S' => Self::Dir {
                    pos: pos.move_direction(Cardinal::South, amount as i64).unwrap(),
                    dir,
                },
                'E' => Self::Dir {
                    pos: pos.move_direction(Cardinal::East, amount as i64).unwrap(),
                    dir,
                },
                'W' => Self::Dir {
                    pos: pos.move_direction(Cardinal::West, amount as i64).unwrap(),
                    dir,
                },
                'F' => Self::Dir {
                    pos: pos.move_direction(dir, amount as i64).unwrap(),
                    dir,
                },
                'L' => Self::Dir {
                    pos,
                    dir: multi_turn(dir, amount, Cardinal::left),
                },
                'R' => Self::Dir {
                    pos,
                    dir: multi_turn(dir, amount, Cardinal::right),
                },
                c => {
                    panic!("Invalid instruction: {c}");
                }
            },
            ShipState::Waypoint { pos, waypoint } => match action {
                'N' => Self::Waypoint {
                    pos,
                    waypoint: waypoint
                        .move_direction(Cardinal::North, amount as i64)
                        .unwrap(),
                },
                'S' => Self::Waypoint {
                    pos,
                    waypoint: waypoint
                        .move_direction(Cardinal::South, amount as i64)
                        .unwrap(),
                },
                'E' => Self::Waypoint {
                    pos,
                    waypoint: waypoint
                        .move_direction(Cardinal::East, amount as i64)
                        .unwrap(),
                },
                'W' => Self::Waypoint {
                    pos,
                    waypoint: waypoint
                        .move_direction(Cardinal::West, amount as i64)
                        .unwrap(),
                },
                'F' => Self::Waypoint {
                    pos: (0..amount).fold(pos, |acc, _| acc + waypoint),
                    waypoint,
                },
                'L' => Self::Waypoint {
                    pos,
                    waypoint: (0..(amount / 90) % 4)
                        .fold(waypoint, |acc, _| Pos::new2d(acc.y(), -acc.x())),
                },
                'R' => Self::Waypoint {
                    pos,
                    waypoint: (0..(amount / 90) % 4)
                        .fold(waypoint, |acc, _| Pos::new2d(-acc.y(), acc.x())),
                },
                c => {
                    panic!("Invalid instruction: {c}");
                }
            },
        }
    }

    fn distance(&self) -> Output {
        let pos = match self {
            ShipState::Dir { pos, dir: _ } => *pos,
            ShipState::Waypoint { pos, waypoint: _ } => *pos,
        };
        pos.manhattan_distance(&Pos::origin())
    }
}

fn multi_turn<F>(dir: Cardinal, amount: Output, turn: F) -> Cardinal
where
    F: Fn(&Cardinal) -> Cardinal,
{
    (0..(amount / 90) % 4).fold(dir, |acc, _| turn(&acc))
}

fn parse_input(input: &str) -> Input {
    input
        .lines()
        .map(|line| {
            let (action, amount) = line.split_at(1);
            let action = action.chars().next().unwrap();
            let amount = amount.parse().unwrap();
            (action, amount)
        })
        .collect()
}

fn solve(instructions: &Input, initial_ship_state: ShipState) -> Output {
    instructions
        .iter()
        .fold(initial_ship_state, |state, &instruction| {
            state.next_state(instruction)
        })
        .distance()
}

fn part1(instructions: &Input) -> Output {
    solve(
        instructions,
        ShipState::Dir {
            pos: Pos::origin(),
            dir: Cardinal::East,
        },
    )
}

fn part2(instructions: &Input) -> Output {
    solve(
        instructions,
        ShipState::Waypoint {
            pos: Pos::origin(),
            waypoint: Pos::new2d(10, -1),
        },
    )
}

#[test]
fn default() {
    let input = get_input(20, 12).unwrap();
    let input = parse_input(&input);
    assert_eq!(2280, part1(&input));
    assert_eq!(38693, part2(&input));
}

// Input parsed (41μs)
// 1. 2280 (15μs)
// 2. 38693 (11μs)
// Total: 71μs
